src/Expander/Primitives.hs

```diff
@@ -529,41 +529,73 @@ arrowType :: TypePrim
 arrowType = (implT, implP)
   where
     implT k dest stx = do
-      equateKinds stx KStar k
-      Stx _ _ (_, arg, ret) <- mustHaveEntries stx
-      argDest <- scheduleType KStar arg
-      retDest <- scheduleType KStar ret
-      linkType dest (tFun1 argDest retDest)
+      Stx _ _ oneToThreeEntries <- mustHaveEntries stx
+      case oneToThreeEntries of
+        Left (Identity _) -> do
+          equateKinds stx (kFun [KStar, KStar] KStar) k
+          linkType dest $ TyF TFun []
+        Right (Left (_, arg)) -> do
+          equateKinds stx (kFun [KStar] KStar) k
+          argDest <- scheduleType KStar arg
+          linkType dest $ TyF TFun [argDest]
+        Right (Right (_, arg, ret)) -> do
+          equateKinds stx KStar k
+          argDest <- scheduleType KStar arg
+          retDest <- scheduleType KStar ret
+          linkType dest $ tFun1 argDest retDest
     implP dest stx = do
-      Stx _ _ (_, arg, ret) <- mustHaveEntries stx
-      (sc1, n1, x1) <- prepareVar arg
-      (sc2, n2, x2) <- prepareVar ret
-      sch <- trivialScheme tType
-      linkTypePattern dest
-        (TypePattern (tFun1 (n1, x1) (n2, x2)))
-        [(sc1, n1, x1, sch), (sc2, n2, x2, sch)]
+      Stx _ _ oneToThreeEntries <- mustHaveEntries stx
+      case oneToThreeEntries of
+        Left (Identity _) -> do
+          linkTypePattern dest
+            (TypePattern (TyF TFun []))
+            []
+        Right (Left (_, arg)) -> do
+          (sc1, n1, x1) <- prepareVar arg
+          sch <- trivialScheme tType
+          linkTypePattern dest
+            (TypePattern (TyF TFun [(n1, x1)]))
+            [(sc1, n1, x1, sch)]
+        Right (Right (_, arg, ret)) -> do
+          (sc1, n1, x1) <- prepareVar arg
+          (sc2, n2, x2) <- prepareVar ret
+          sch <- trivialScheme tType
+          linkTypePattern dest
+            (TypePattern (tFun1 (n1, x1) (n2, x2)))
+            [(sc1, n1, x1, sch), (sc2, n2, x2, sch)]
 
 macroType :: TypePrim
-macroType = unaryType (\a -> tMacro a)
+macroType = unaryType TMacro
 
 ioType :: TypePrim
-ioType = unaryType tIO
+ioType = unaryType TIO
 
-unaryType :: (forall a . a -> TyF a) -> TypePrim
+unaryType :: TypeConstructor -> TypePrim
```

Comment:
This is exactly what we'd expect it to have with proper kinds. But I wonder whether we should have `unaryType` instead of something like `typeConstructor :: TypeConstructor -> Kind -> TypePrim`. This would move some errors from being syntax errors to being kind errors, but that seems reasonable enough to me. Am I missing something?

Reply:
Then the same implementation could be used for declared datatypes as for builtins like IO, increasing the robustness of the system.

Reply:
Unifying our series of specialized `plainType`, `unaryType` and `binaryType` functions into a single `typeConstructor` function does seem like a good idea, thanks!

Reply:
That one is a bit more involved, I'll do it in a separate PR

Reply:
Created #195 to track this, so I don't forget